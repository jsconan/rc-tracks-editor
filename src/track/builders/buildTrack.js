/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { Counter, Vector2D } from '../../core/models';
import { TilesList } from '../models';

/**
 * Process a list of tiles for rendering a track, computing the coordinates of each tile.
 * @param {TilesList} list - The list of tiles from which build the track.
 * @param {object} config - A set of config options.
 * @param {number} [config.startX] - The X-coordinate of the first tile.
 * @param {number} [config.startY] - The Y-coordinate of the first tile.
 * @param {number} [config.startAngle] - The rotation angle of the first tile.
 * @param {number} [config.hPadding] - An horizontal padding added around the track.
 * @param {number} [config.vPadding] - A vertical padding added around the tracks.
 * @returns {trackCoord}
 * @throws {TypeError} - If the given list is not a valid instance of TileList.
 */
export default (list, { startX = 0, startY = 0, startAngle = 0, hPadding = 0, vPadding = 0 } = {}) => {
    TilesList.validateInstance(list);

    const topLeft = new Vector2D();
    const bottomRight = new Vector2D();
    let inputX = startX;
    let inputY = startY;
    let inputAngle = startAngle;

    const stats = new Counter();
    const tiles = list.tiles.map(model => {
        const bounds = model.getBoundingRect(inputX, inputY, inputAngle);
        const { x, y, angle } = bounds.input;
        const { id, type, direction, ratio } = model;

        stats.increment(model.modelId);

        inputX = bounds.output.x;
        inputY = bounds.output.y;
        inputAngle = bounds.output.angle;

        topLeft.x = Math.min(topLeft.x, bounds.x);
        topLeft.y = Math.min(topLeft.y, bounds.y);

        bottomRight.x = Math.max(bottomRight.x, bounds.x + bounds.width);
        bottomRight.y = Math.max(bottomRight.y, bounds.y + bounds.height);

        return { id, type, direction, ratio, x, y, angle, model };
    });

    const { x, y } = topLeft.subCoord(hPadding, vPadding);
    const { x: width, y: height } = bottomRight.sub(topLeft).addCoord(hPadding * 2, vPadding * 2);

    return { x, y, width, height, tiles, stats };
};

/**
 * @typedef {object} trackCoord - Represents a track ready to be rendered.
 * @property {number} x - The left coordinate of the track.
 * @property {number} y - The top coordinate of the track.
 * @property {number} width - The width of the track.
 * @property {number} height - The height of the track.
 * @property {tileCoord[]} tiles - The list of tiles.
 * @property {Counter} stats - An object listing the stats for the track.
 */

/**
 * @typedef {object} tileCoord - Represents a positioned tile.
 * @property {string} id - The unique identifier of the tile.
 * @property {string} type - The type of tile.
 * @property {string} direction - The direction of the tile.
 * @property {number} ratio - The size ratio of the tile.
 * @property {number} x - The left coordinate of the tile.
 * @property {number} y - The top coordinate of the tile.
 * @property {number} angle - The rotation angle of the tile.
 * @property {TileModel} model - A reference to the tile model.
 */

/**
 * @typedef {import('../models').TileModel} TileModel
 */
