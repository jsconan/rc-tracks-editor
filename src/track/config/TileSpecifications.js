/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Gives the specifications for the track tiles with respect to the initial constraints.
 */
export class TileSpecifications {
    /**
     * Defines the specifications for the track tiles from the given constraints.
     * @param {number} laneWidth - The width of the track lane (the distance between the barriers).
     * @param {number} barrierWidth - The width of the barriers.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @param {number} maxRatio - The maximum value for size ratios.
     */
    constructor(laneWidth = 20, barrierWidth = 1, barrierChunks = 4, maxRatio = 4) {
        this.setLaneWidth(laneWidth);
        this.setBarrierWidth(barrierWidth);
        this.setBarrierChunks(barrierChunks);
        this.setMaxRatio(maxRatio);
    }

    /**
     * The length of a tile with respect to the initial constraints.
     * @type {number}
     */
    get length() {
        return this.laneWidth * 1.25 + this.barrierWidth * 2;
    }

    /**
     * The width of a tile with respect to the initial constraints.
     * @type {number}
     */
    get width() {
        return this.laneWidth + this.barrierWidth * 2;
    }

    /**
     * The length of a barrier chunk with respect to the initial constraints.
     * @type {number}
     */
    get barrierLength() {
        return this.length / this.barrierChunks;
    }

    /**
     * The padding around the width of the tile to align with its length.
     * Note: tiles must fit in a square so that they can be flipped and rotated indifferently.
     * @type {number}
     */
    get padding() {
        return this.laneWidth / 8;
    }

    /**
     * Sets the width of the track lane (the distance between the barriers).
     * @param {number} laneWidth - The width of the track lane.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setLaneWidth(laneWidth) {
        this.laneWidth = Math.abs(laneWidth);

        return this;
    }

    /**
     * Sets the width of the barriers.
     * @param {number} barrierWidth - The width of the barriers.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierWidth(barrierWidth) {
        this.barrierWidth = Math.abs(barrierWidth);

        return this;
    }

    /**
     * Sets the number of barrier chunks per tile.
     * @param {number} barrierChunks - The number of barrier chunks per tile.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setBarrierChunks(barrierChunks) {
        this.barrierChunks = Math.abs(Math.round(barrierChunks) || 1);

        return this;
    }

    /**
     * Sets the maximum value for size ratios.
     * @param {number} maxRatio - The maximum value for size ratios.
     * @returns {TileSpecifications} - Chains the instance.
     */
    setMaxRatio(maxRatio) {
        this.maxRatio = Math.abs(Math.round(maxRatio) || 1);

        return this;
    }
}
