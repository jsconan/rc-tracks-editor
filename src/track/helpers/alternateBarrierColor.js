/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { alternate } from '../../core/helpers';

/**
 * A list of colors for the barriers.
 * @type {string[]}
 */
const barrierColors = ['even', 'odd'];

/**
 * Returns the color for a barrier element at the given position.
 * @param {number} i - The position of the barrier element.
 * @returns {string} - The color for the barrier element.
 */
export default i => alternate(i, barrierColors);
