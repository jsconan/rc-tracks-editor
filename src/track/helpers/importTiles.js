/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { TileList } from '../../tile/models';
import createTile from './createTile.js';

/**
 * Imports a list of tiles from exported data.
 * @param {TileList} list - The list in which import the tiles.
 * @param {tileExport[]} data - An iterable object containing exported tiles.
 * @returns {TileList} - Returns the list.
 */
export default (list, data) => {
    TileList.validateInstance(list);

    if (!data || !data[Symbol.iterator]) {
        return list;
    }

    const dataIterator = data[Symbol.iterator]();
    const loadIterator = {
        next() {
            const next = dataIterator.next();

            if (next.done) {
                return next;
            }

            const { type, direction, ratio } = next.value || {};
            const tile = createTile(list.specs, type, direction, ratio);

            return { done: false, value: tile };
        },

        [Symbol.iterator]() {
            return this;
        }
    };

    list.load(loadIterator);

    return list;
};

/**
 * @typedef {import('../../tile/models/TileModel.js').tileExport} tileExport
 */
