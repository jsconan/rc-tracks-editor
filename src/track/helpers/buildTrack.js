/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { validateAPI } from '../../core/helpers';
import { Vector2D } from '../../core/models';

/**
 * Process a list of tiles for rendering a track, computing the coordinates of each tile.
 * @param {*} list - The list of tiles from which build the track. It must implement a map method.
 * @param {object} config - A set of config options.
 * @param {number} [config.startX] - The X-coordinate of the first tile.
 * @param {number} [config.startY] - The Y-coordinate of the first tile.
 * @param {number} [config.startAngle] - The rotation angle of the first tile.
 * @param {number} [config.hPadding] - An horizontal padding added around the track.
 * @param {number} [config.vPadding] - A vertical padding added around the tracks.
 * @returns {trackCoord} - Returns the list of coordinates.
 * @throws {TypeError} - If the given list does not implement the map method.
 */
export default (list, { startX = 0, startY = 0, startAngle = 0, hPadding = 0, vPadding = 0 } = {}) => {
    validateAPI(list, ['map']);

    const topLeft = new Vector2D();
    const bottomRight = new Vector2D();
    let inputX = startX;
    let inputY = startY;
    let inputAngle = startAngle;

    const tiles = list.map(model => {
        const rect = model.getBoundingRect(inputX, inputY, inputAngle);
        const { x, y, angle } = rect.input;
        const { id, type, direction, ratio } = model;

        inputX = rect.output.x;
        inputY = rect.output.y;
        inputAngle = rect.output.angle;

        topLeft.x = Math.min(topLeft.x, rect.x);
        topLeft.y = Math.min(topLeft.y, rect.y);

        bottomRight.x = Math.max(bottomRight.x, rect.x + rect.width);
        bottomRight.y = Math.max(bottomRight.y, rect.y + rect.height);

        return { id, type, direction, ratio, x, y, angle, rect, model };
    });

    const { x, y } = topLeft.subCoord(hPadding, vPadding);
    const { x: width, y: height } = bottomRight.sub(topLeft).addCoord(hPadding * 2, vPadding * 2);

    return { x, y, width, height, tiles };
};

/**
 * @typedef {object} trackCoord - Represents a track ready to be rendered.
 * @property {number} x - The left coordinate of the track.
 * @property {number} y - The top coordinate of the track.
 * @property {number} width - The width of the track.
 * @property {number} height - The height of the track.
 * @property {tileCoord[]} tiles - The list of tiles.
 */

/**
 * @typedef {object} tileCoord - Represents a positioned tile.
 * @property {string} id - The unique identifier of the tile.
 * @property {string} type - The type of tile.
 * @property {string} direction - The direction of the tile.
 * @property {number} ratio - The size ratio of the tile.
 * @property {number} x - The left coordinate of the tile.
 * @property {number} y - The top coordinate of the tile.
 * @property {number} angle - The rotation angle of the tile.
 * @property {TileModel} model - A reference to the tile model.
 */

/**
 * @typedef {import('../../tile/models').TileModel} TileModel
 */
