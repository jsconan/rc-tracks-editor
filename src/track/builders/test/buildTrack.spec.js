/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import buildTrack from '../buildTrack.js';

import { CURVED_TILE_TYPE, STRAIGHT_TILE_TYPE, TILE_DIRECTION_RIGHT } from '../../helpers';
import { TileSpecifications } from '../../config';
import { TileList } from '../../models';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const specs = new TileSpecifications(laneWidth, barrierWidth, barrierChunks);
const list = new TileList(specs);
list.import([
    { type: STRAIGHT_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: CURVED_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: STRAIGHT_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: CURVED_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: STRAIGHT_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: CURVED_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: STRAIGHT_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 },
    { type: CURVED_TILE_TYPE, direction: TILE_DIRECTION_RIGHT, ratio: 1 }
]);

describe('buildTrack', () => {
    it('is a function', () => {
        expect(buildTrack).toEqual(expect.any(Function));
    });

    it('throws error when trying to use a wrong list of tiles', () => {
        expect(() => buildTrack([])).toThrow('The object must be an instance of TileList!');
    });

    describe('process a list of tiles for rendering a track', () => {
        it('using default position', () => {
            expect(buildTrack(list)).toMatchSnapshot();
        });

        it.each([
            {},
            { startX: 100, startY: 100 },
            { startAngle: 45 },
            { startX: 100, startY: 100, startAngle: 45 },
            { hPadding: 10, vPadding: 20 },
            { startX: 100, startY: 100, startAngle: 45, hPadding: 10, vPadding: 20 }
        ])('using config options %s', config => {
            expect(buildTrack(list, config)).toMatchSnapshot();
        });
    });
});
