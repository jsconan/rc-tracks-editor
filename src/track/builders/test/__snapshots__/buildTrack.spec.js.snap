// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTrack process a list of tiles for rendering a track using config options { startAngle: 45 } 1`] = `
Object {
  "height": 355.56349186104046,
  "stats": Counter {
    "counters": Map {
      "straight-tile-1" => 4,
      "curved-tile-1" => 4,
    },
    "notify": [Function],
    "subscribe": [Function],
  },
  "tiles": Array [
    Object {
      "angle": 45,
      "id": "id-0",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-0",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 0,
      "y": 0,
    },
    Object {
      "angle": 45,
      "id": "id-1",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-1",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -77.78174593052022,
      "y": 77.78174593052023,
    },
    Object {
      "angle": 135,
      "id": "id-2",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-2",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -155.56349186104046,
      "y": 77.78174593052023,
    },
    Object {
      "angle": 135,
      "id": "id-3",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-3",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -233.3452377915607,
      "y": 1.4210854715202004e-14,
    },
    Object {
      "angle": 225,
      "id": "id-4",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-4",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -233.3452377915607,
      "y": -77.78174593052022,
    },
    Object {
      "angle": 225,
      "id": "id-5",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-5",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -155.56349186104046,
      "y": -155.56349186104046,
    },
    Object {
      "angle": 315,
      "id": "id-6",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-6",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -77.78174593052023,
      "y": -155.56349186104046,
    },
    Object {
      "angle": 315,
      "id": "id-7",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-7",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 1.4210854715202004e-14,
      "y": -77.78174593052026,
    },
  ],
  "width": 355.56349186104046,
  "x": -294.45436482630055,
  "y": -216.67261889578035,
}
`;

exports[`buildTrack process a list of tiles for rendering a track using config options { startX: 100, startY: 100 } 1`] = `
Object {
  "height": 310,
  "stats": Counter {
    "counters": Map {
      "straight-tile-1" => 4,
      "curved-tile-1" => 4,
    },
    "notify": [Function],
    "subscribe": [Function],
  },
  "tiles": Array [
    Object {
      "angle": 0,
      "id": "id-0",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-0",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 100,
      "y": 100,
    },
    Object {
      "angle": 0,
      "id": "id-1",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-1",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 100,
      "y": 210,
    },
    Object {
      "angle": 90,
      "id": "id-2",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-2",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 45,
      "y": 265,
    },
    Object {
      "angle": 90,
      "id": "id-3",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-3",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -65,
      "y": 265,
    },
    Object {
      "angle": 180,
      "id": "id-4",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-4",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -120,
      "y": 210,
    },
    Object {
      "angle": 180,
      "id": "id-5",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-5",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -120.00000000000001,
      "y": 100,
    },
    Object {
      "angle": 270,
      "id": "id-6",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-6",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -65.00000000000003,
      "y": 44.99999999999999,
    },
    Object {
      "angle": 270,
      "id": "id-7",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-7",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 44.99999999999997,
      "y": 44.99999999999997,
    },
  ],
  "width": 310,
  "x": -165.00000000000003,
  "y": -2.842170943040401e-14,
}
`;

exports[`buildTrack process a list of tiles for rendering a track using config options { startX: 100, startY: 100, startAngle: 45 } 1`] = `
Object {
  "height": 355.56349186104046,
  "stats": Counter {
    "counters": Map {
      "straight-tile-1" => 4,
      "curved-tile-1" => 4,
    },
    "notify": [Function],
    "subscribe": [Function],
  },
  "tiles": Array [
    Object {
      "angle": 45,
      "id": "id-0",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-0",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 100,
      "y": 100,
    },
    Object {
      "angle": 45,
      "id": "id-1",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-1",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 22.218254069479784,
      "y": 177.78174593052023,
    },
    Object {
      "angle": 135,
      "id": "id-2",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-2",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -55.56349186104045,
      "y": 177.78174593052023,
    },
    Object {
      "angle": 135,
      "id": "id-3",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-3",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -133.3452377915607,
      "y": 100.00000000000001,
    },
    Object {
      "angle": 225,
      "id": "id-4",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-4",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -133.3452377915607,
      "y": 22.218254069479798,
    },
    Object {
      "angle": 225,
      "id": "id-5",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-5",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -55.563491861040475,
      "y": -55.56349186104045,
    },
    Object {
      "angle": 315,
      "id": "id-6",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-6",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 22.218254069479755,
      "y": -55.56349186104046,
    },
    Object {
      "angle": 315,
      "id": "id-7",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-7",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 100,
      "y": 22.21825406947974,
    },
  ],
  "width": 355.56349186104046,
  "x": -194.45436482630058,
  "y": -116.67261889578035,
}
`;

exports[`buildTrack process a list of tiles for rendering a track using config options {} 1`] = `
Object {
  "height": 310,
  "stats": Counter {
    "counters": Map {
      "straight-tile-1" => 4,
      "curved-tile-1" => 4,
    },
    "notify": [Function],
    "subscribe": [Function],
  },
  "tiles": Array [
    Object {
      "angle": 0,
      "id": "id-0",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-0",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 0,
      "y": 0,
    },
    Object {
      "angle": 0,
      "id": "id-1",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-1",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 0,
      "y": 110,
    },
    Object {
      "angle": 90,
      "id": "id-2",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-2",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -55,
      "y": 165,
    },
    Object {
      "angle": 90,
      "id": "id-3",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-3",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -165,
      "y": 165,
    },
    Object {
      "angle": 180,
      "id": "id-4",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-4",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -220,
      "y": 110,
    },
    Object {
      "angle": 180,
      "id": "id-5",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-5",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -220,
      "y": 0,
    },
    Object {
      "angle": 270,
      "id": "id-6",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-6",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -165,
      "y": -55.00000000000001,
    },
    Object {
      "angle": 270,
      "id": "id-7",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-7",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -55,
      "y": -55.00000000000003,
    },
  ],
  "width": 310,
  "x": -265,
  "y": -100.00000000000003,
}
`;

exports[`buildTrack process a list of tiles for rendering a track using default position 1`] = `
Object {
  "height": 310,
  "stats": Counter {
    "counters": Map {
      "straight-tile-1" => 4,
      "curved-tile-1" => 4,
    },
    "notify": [Function],
    "subscribe": [Function],
  },
  "tiles": Array [
    Object {
      "angle": 0,
      "id": "id-0",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-0",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 0,
      "y": 0,
    },
    Object {
      "angle": 0,
      "id": "id-1",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-1",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": 0,
      "y": 110,
    },
    Object {
      "angle": 90,
      "id": "id-2",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-2",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -55,
      "y": 165,
    },
    Object {
      "angle": 90,
      "id": "id-3",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-3",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -165,
      "y": 165,
    },
    Object {
      "angle": 180,
      "id": "id-4",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-4",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -220,
      "y": 110,
    },
    Object {
      "angle": 180,
      "id": "id-5",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-5",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -220,
      "y": 0,
    },
    Object {
      "angle": 270,
      "id": "id-6",
      "model": StraightTileModel {
        "direction": "right",
        "id": "id-6",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -165,
      "y": -55.00000000000001,
    },
    Object {
      "angle": 270,
      "id": "id-7",
      "model": CurvedTileModel {
        "direction": "right",
        "id": "id-7",
        "ratio": 1,
        "specs": TileSpecifications {
          "barrierChunks": 4,
          "barrierWidth": 5,
          "laneWidth": 80,
          "maxRatio": 4,
        },
      },
      "x": -55,
      "y": -55.00000000000003,
    },
  ],
  "width": 310,
  "x": -265,
  "y": -100.00000000000003,
}
`;
